### PR TITLE
Linear regression implementation.

### PR DESCRIPTION
**Pull request by param087**
#4

### CODE REVIEW

File: Sources/swiftML/Helpers.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  
+/// Return the minkowski distance based on value of p.
+///
+/// - Parameters
+///   - a: Input tensor to find distance.
+///   - b: Input tensor to find distance.
+///   - p: The distance metric to use for the tree, default set to 2.
+/// - Returns: Minkowski distance based on value of p.
+public func minkowskiDistance(_ a: Tensor<Float>, _ b: Tensor<Float>, _ p: Int ) -> Tensor<Float> {
```
Comment (dan-zheng) [inline Sources/swiftML/Helpers.swift]: How about including the argument label `p`, to distinguish it from the unlabeled arguments?
```suggestion
public func minkowskiDistance(_ a: Tensor<Float>, _ b: Tensor<Float>, p: Int) -> Tensor<Float> {
```

File: Sources/swiftML/Helpers.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        let signs = Raw.sign(rowValueForSign)
+        U = U * signs
+        V = V * signs.reshaped(to: [u.shape[1], 1])
+    }
+    
+    return (U, V)
+}
+
+  
+/// Return the minkowski distance based on value of p.
```
Comment (dan-zheng) [inline Sources/swiftML/Helpers.swift]: Documentation comment suggestions:
```suggestion
/// Returns the Minkowski distance between two tensors for the given distance metric `p`.
/// - Parameters:
///   - a: The first tensor.
///   - b: The second tensor.
///   - p: The order of the norm of the difference: `||a - b||_p`.
```

I don't feel a `/// - Returns` comment is necessary because the primary doc comment starts with `/// Returns`.

File: Sources/swiftML/Helpers.swift (hunk 1 above)
Comment (dan-zheng) [inline Sources/swiftML/Helpers.swift]: What inputs are `minkowskiDistance` and `euclideanDistance` designed to take?
(1D tensors? tensors with the same shapes?)

Currently, without preconditions, the functions will work with tensors of different shapes, performing broadcasting. Consider an appropriate precondition, such as:
```swift
  precondition(a.shape == b.shape)
```

Comment (param087) [reply Sources/swiftML/Helpers.swift]: That seems more appropriate in the case of reusability and understanding.

File: Sources/swiftML/Helpers.swift (hunk 2 above)
Comment (param087) [reply Sources/swiftML/Helpers.swift]: I am thinking of specifying the data type/tensor structure in ` /// - Parameters` for type safety. What do you think about that?

File: Sources/swiftML/Helpers.swift (hunk 1 above)
Comment (param087) [reply Sources/swiftML/Helpers.swift]: Input to `minkowskiDistance` and `euclideanDistance` are 1D tensor of shape [n] . I will add appropriate precondition.

File: Sources/swiftML/Helpers.swift (hunk 2 above)
Comment (param087) [reply Sources/swiftML/Helpers.swift]: hey @dan-zheng,
For the documentation point of view, I was thinking of using [https://swift.org/documentation/api-design-guidelines/](https://swift.org/documentation/api-design-guidelines/)  and ([https://developer.apple.com/library/archive/documentat1ion/Xcode/Reference/xcode_markup_formatting_ref/index.html#//apple_ref/doc/uid/TP40016497-CH2-SW1](https://developer.apple.com/library/archive/documentation/Xcode/Reference/xcode_markup_formatting_ref/index.html#//apple_ref/doc/uid/TP40016497-CH2-SW1).) Is it ok? Or you have some specific guideline in mind?

Comment (rxwei) [reply Sources/swiftML/Helpers.swift]: You don't need to specify types in `- Parameter:` comments, since Swift is already statically and strongly typed.

Comment (param087) [reply Sources/swiftML/Helpers.swift]: yeah, Before seeing documentation I was thinking of that but now I don't.

File: Sources/swiftML/Helpers.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+///   - u: The tensor containing of left singular vectors for each matrix.
+///   - v: The tensor containing of right singular vectors for each matrix.
+/// - Returns: Sign correction to ensure deterministic output.
+public func svdFlip(
+    u: Tensor<Double>,
+    v: Tensor<Double>,
+    uBasedDecision: Bool = true
+) -> (Tensor<Double>, Tensor<Double>) {
+  
+    var U = u
```
Comment (dan-zheng) [inline Sources/swiftML/Helpers.swift]: Please create mutable local variables named `u` and `v` instead of `U` and `V`. Swift supports this kind of aliasing.
```suggestion
    var u = u
    var v = v
```

**Comment by dan-zheng** [comment]
I pushed some commits with changes.

Notable changes:
* Use consistent spelling in documentation comments and precondition messages.
* Change helper functions to be generic.
* Refactor `func matrixPseudoInverse` into a computed property `Tensor.pseudoinverse`, which is more idiomatic.

@param087: could you please review the changes to see if they make sense?

I saw that many PRs depend on `Helpers.swift` - could you please open a separate PR containing just `Helpers.swift`? Let's try to merge that PR ASAP to unblock your progress.

I have some more feedback regarding `LinearRegression.swift`, commenting soon.

File: Sources/swiftML/LinearRegression.swift [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        precondition(labels.shape[1] == 1, "Labels must have single target feature.")
+        precondition(data.shape[0] == labels.shape[0],
+                     "Data and labels must have the same sample count.")
+
+        var data: Tensor<Float> = data
+        if self.fitIntercept {
+            let ones = Tensor<Float>(ones: [data.shape[0], 1])
+            data = ones.concatenated(with: data, alongAxis: -1)
+        }
+        if self.gradientDescent == false {
```
Comment (dan-zheng) [inline Sources/swiftML/LinearRegression.swift]: It seems that `class LinearRegression` currently has two different "strategies":
1. Gradient descent: which uses `fitIntercept`, `weights`, `learningRate`, `iterations`.
2. Singular value decomposition: which uses `fitIntercept`, `weights`.

Right now, the strategy is controlled by a flag `var gradientDescent: Bool`, but this is not particularly safe. `gradientDescent` may be `false`, but `learningRate` and `iterations` can still be accessed.

---

To make the design more safe, consider one of the following:
* Create a `protocol LinearRegression` and implement two separate conforming types: `class GradientDescentLinearRegression` and `class SingleValueDecompositionLinearRegression`.
  * `protocol LinearRegression` defines requirement `func fit(data:labels:)` and provides default implementations for `func prediction(for:)` and `func score(data:labels:)` in an extension.
  * `class GradientDescentLinearRegression` and `class SingleValueDecompositionLinearRegression` conform to `protocol LinearRegression`, implement `func fit`, and define just the stored properties that they need. This way, `SingleValueDecompositionLinearRegression` won't be able to access `learningRate` or `iterationCount`.
* Keep a single `class LinearRegression`, but define a nested enum representing the "strategy".
  ```swift
  public class LinearRegresion {
      /// A linear regression strategy.
      enum Strategy {
          /// Gradient descent.
          case gradientDescent(learningRate: Float, iterationCount: Int)

          /// Singular value decomposition.
          case singularValueDecomposition
      }
      var strategy: Strategy // replaces `var gradientDescent: Bool`
      ...
  }
  ```
  * Then, create two factory methods: `static func gradientDescent(learningRate: Float, iterationCount: Int) -> LinearRegression` and `static func singleValueDecomposition() -> LinearRegression`. These will be used to construct instances of `LinearRegression`.

cc @rxwei

Comment (param087) [reply Sources/swiftML/LinearRegression.swift]: @dan-zheng - I have started working on it.

Comment (param087) [reply Sources/swiftML/LinearRegression.swift]: @dan-zheng - I have a quick question for you. What level of documentation should in protocol and implementation?

Comment (dan-zheng) [reply Sources/swiftML/LinearRegression.swift]: A 1-2 sentence summary per type declaration/function should be good!

For types/functions that implement well-known concepts (e.g. linear regression, matrix pseudoinverse, Minkowski distance), consider adding a reference link (e.g. to Wikipedia).

You can look at [tensorflow/swift-apis](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Loss.swift) as a reference for documentation: see [optimizer](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Optimizers/MomentumBased.swift) documentation.
